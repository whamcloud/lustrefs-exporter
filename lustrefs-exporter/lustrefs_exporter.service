[Unit]
Description=Prometheus exporter for Lustre filesystem
Documentation=https://github.com/whamcloud/lustrefs-exporter

[Service]
Environment=RUST_LOG=info
Restart=on-failure
ExecStart=/usr/bin/lustrefs_exporter
MemoryHigh=1750M
MemoryMax=2G

[Install]
WantedBy=multi-user.target
