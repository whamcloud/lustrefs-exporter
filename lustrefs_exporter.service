[Unit]
Description=Prometheus exporter for Lustre filesystem
Documentation=https://github.com/whamcloud/lustrefs-exporter

[Service]
Restart=on-failure
User=prometheus
ExecStart=/usr/bin/lustrefs_exporter

[Install]
WantedBy=multi-user.target

