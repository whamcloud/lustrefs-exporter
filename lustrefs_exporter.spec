Name:           lustrefs_exporter
Version:        0.2.0
Release:        1%{?dist}
Summary:        prometheus exporter for lustre
License:        MIT

Requires(pre): shadow-utils

%description
Prometheus exporter for the Lustre filesystem

%global debug_package %{nil}

%prep

%build
cargo build --release

%install
install -v -d %{buildroot}%{_bindir}
install -v -d %{buildroot}%{_unitdir}
install -v -m 0644 lustrefs_exporter.service %{buildroot}%{_unitdir}
install -v target/release/lustrefs-exporter %{buildroot}%{_bindir}
%{__ln_s} lustrefs-exporter %{buildroot}%{_bindir}/lustrefs_exporter

%files
%{_bindir}/lustrefs-exporter
%{_bindir}/lustrefs_exporter
%{_unitdir}/lustrefs_exporter.service

%pre
getent group prometheus >/dev/null || groupadd -r prometheus
getent passwd prometheus >/dev/null || \
  useradd -r -g prometheus -s /sbin/nologin \
          -c "Prometheus services" prometheus

%post
%systemd_post %{name}.service

%preun
%systemd_preun %{name}.service

%postun
%systemd_postun %{name}.service
