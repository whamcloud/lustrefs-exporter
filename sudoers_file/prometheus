prometheus ALL=(ALL) NOPASSWD: /usr/sbin/lnetctl net show -v 4, /usr/sbin/lnetctl stats show
