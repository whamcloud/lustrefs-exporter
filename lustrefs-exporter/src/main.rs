// Copyright (c) 2024 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use axum::{
    body::Body,
    error_handling::HandleErrorLayer,
    extract::Query,
    http::{self, HeaderValue, StatusCode},
    response::{IntoResponse, Response},
    routing::get,
    BoxError, Router,
};
use clap::Parser;
use lustre_collector::{parse_lctl_output, parse_lnetctl_output, parse_lnetctl_stats, parser};
use lustrefs_exporter::{
    jobstats::opentelemetry::OpenTelemetryMetricsJobstats,
    openmetrics::{self, OpenTelemetryMetrics},
    Error,
};
use opentelemetry::{
    global,
    metrics::MeterProvider,
};
use opentelemetry_sdk::{metrics::SdkMeterProvider, Resource};
use prometheus::{Encoder as _, Registry, TextEncoder};
use serde::Deserialize;
use std::{
    borrow::Cow,
    convert::Infallible,
    io::{self, BufRead, BufReader},
    net::SocketAddr,
    sync::Arc,
};
use tokio::process::Command;
use tower::ServiceBuilder;

const LUSTREFS_EXPORTER_PORT: &str = "32221";

#[derive(Debug, Parser)]
pub struct CommandOpts {
    /// Port that exporter will listen to
    #[clap(short, long, env = "LUSTREFS_EXPORTER_PORT", default_value = LUSTREFS_EXPORTER_PORT)]
    pub port: u16,
}

async fn handle_error(error: BoxError) -> impl IntoResponse {
    if error.is::<tower::timeout::error::Elapsed>() {
        return (StatusCode::REQUEST_TIMEOUT, Cow::from("request timed out"));
    }

    if error.is::<tower::load_shed::error::Overloaded>() {
        return (
            StatusCode::SERVICE_UNAVAILABLE,
            Cow::from("service is overloaded, try again later"),
        );
    }

    (
        StatusCode::INTERNAL_SERVER_ERROR,
        Cow::from(format!("Unhandled internal error: {error}")),
    )
}

#[derive(Debug, Deserialize)]
struct Params {
    // Only enable jobstats if "jobstats=true"
    #[serde(default)]
    jobstats: bool,
}

pub fn init_opentelemetry() -> Result<
    (opentelemetry_sdk::metrics::SdkMeterProvider, Registry),
    opentelemetry_sdk::metrics::MetricError,
> {
    // Build the Prometheus exporter.
    // The metrics will be exposed in the Prometheus format.
    let registry = Registry::new();
    let prometheus_exporter = opentelemetry_prometheus::exporter()
        .with_registry(registry.clone())
        .without_counter_suffixes()
        .build()?;

    let provider = SdkMeterProvider::builder()
        .with_reader(prometheus_exporter)
        .with_resource(
            Resource::builder()
                .with_service_name("lustrefs-exporter")
                .build(),
        )
        .build();

    
        // Set the global MeterProvider to the one created above.
    // This will make all meters created with `global::meter()` use the above MeterProvider.
    global::set_meter_provider(provider.clone());

    Ok((provider, registry))
}

#[tokio::main]
async fn main() -> Result<(), Error> {
    tracing_subscriber::fmt::init();

    let opts = CommandOpts::parse();

    let addr = SocketAddr::from(([0, 0, 0, 0], opts.port));

    tracing::info!("Listening on http://{addr}/metrics");

    let listener = tokio::net::TcpListener::bind(("0.0.0.0", opts.port)).await?;

    let load_shedder = ServiceBuilder::new()
        .layer(HandleErrorLayer::new(handle_error))
        .load_shed()
        .concurrency_limit(10); // Max 10 concurrent scrape

    let app = Router::new()
        .route("/metrics", get(scrape))
        .layer(load_shedder);

    axum::serve(listener, app).await?;

    Ok(())
}

async fn scrape(Query(params): Query<Params>) -> Result<Response<Body>, Error> {
    let (provider, registry) = init_opentelemetry()?;

    let meter = provider.meter("lustre");
    let jobstats = if params.jobstats {
        let child = tokio::task::spawn_blocking(move || {
            let child = std::process::Command::new("lctl")
                .arg("get_param")
                .args(["obdfilter.*OST*.job_stats", "mdt.*.job_stats"])
                .stdout(std::process::Stdio::piped())
                .stderr(std::process::Stdio::piped())
                .spawn()?;

            Ok::<_, Error>(child)
        })
        .await?;

        match child {
            Ok(mut child) => {
                let reader = BufReader::with_capacity(
                    128 * 1_024,
                    child.stdout.take().ok_or(io::Error::new(
                        io::ErrorKind::NotFound,
                        "stdout missing for lctl jobstats call.",
                    ))?,
                );

                let reader_stderr = BufReader::new(child.stderr.take().ok_or(io::Error::new(
                    io::ErrorKind::NotFound,
                    "stderr missing for lctl jobstats call.",
                ))?);

                tokio::task::spawn(async move {
                    for line in reader_stderr.lines().map_while(Result::ok) {
                        tracing::debug!("stderr: {}", line);
                    }
                });

                let otel_jobstats = Arc::new(OpenTelemetryMetricsJobstats::new(&meter));

                tokio::task::spawn_blocking(move || {
                    if let Err(e) = child.wait() {
                        tracing::debug!("Unexpected error when waiting for child: {e}");
                    }
                });

                let handle = lustrefs_exporter::jobstats::opentelemetry::jobstats_stream(
                    reader,
                    otel_jobstats.clone(),
                );

                handle.await.unwrap();

                // Encode metrics to string
                let encoder = TextEncoder::new();
                let metric_families = registry.gather();
                let mut output = Vec::new();
                encoder.encode(&metric_families, &mut output).unwrap();

                let output = String::from_utf8_lossy(&output).to_string();
                Some(output)
            }
            Err(e) => {
                tracing::debug!("Error while spawning lctl jobstats: {e}");

                None
            }
        }
    } else {
        None
    };

    let mut output = vec![];

    let lctl = Command::new("lctl")
        .arg("get_param")
        .args(parser::params())
        .kill_on_drop(true)
        .output()
        .await?;

    let mut lctl_output = parse_lctl_output(&lctl.stdout)?;

    output.append(&mut lctl_output);

    let lnetctl = Command::new("lnetctl")
        .args(["net", "show", "-v", "4"])
        .kill_on_drop(true)
        .output()
        .await?;

    let lnetctl_stats = std::str::from_utf8(&lnetctl.stdout)?;
    let mut lnetctl_output = parse_lnetctl_output(lnetctl_stats)?;

    output.append(&mut lnetctl_output);

    let lnetctl_stats_output = Command::new("lnetctl")
        .args(["stats", "show"])
        .kill_on_drop(true)
        .output()
        .await?;

    let mut lnetctl_stats_record =
        parse_lnetctl_stats(std::str::from_utf8(&lnetctl_stats_output.stdout)?)?;

    output.append(&mut lnetctl_stats_record);

    let opentelemetry_metrics = OpenTelemetryMetrics::new(meter.clone());

    let mut lustre_stats = vec![];
    let encoder = TextEncoder::new();
    let metric_families = registry.gather();
    let _encoder_results = encoder.encode(&metric_families, &mut lustre_stats);

    // Build OTEL metrics
    openmetrics::build_lustre_stats(&output, opentelemetry_metrics);

    let mut buffer = vec![];
    let encoder = TextEncoder::new();
    let metric_families = registry.gather();
    encoder.encode(&metric_families, &mut buffer).unwrap();

    let lustre_stats = String::from_utf8_lossy(&buffer).to_string();

    let body = if let Some(stream) = jobstats {
        let merged = tokio_stream::StreamExt::chain(
            tokio_stream::once(Ok::<_, Infallible>(lustre_stats)),
            tokio_stream::once(Ok::<_, Infallible>(stream)),
        );

        Body::from_stream(merged)
    } else {
        tracing::debug!("Jobstats collection disabled");

        Body::from(lustre_stats)
    };

    let mut response_builder = Response::builder().status(StatusCode::OK);

    let headers = response_builder.headers_mut();
    if let Ok(content_type) = encoder.format_type().parse::<HeaderValue>() {
        if let Some(headers) = headers {
            headers.insert(http::header::CONTENT_TYPE, content_type);
        }
    }
    let resp = response_builder.body(body)?;

    Ok(resp)
}

#[cfg(test)]
mod tests {
    use crate::init_opentelemetry;
    use combine::parser::EasyParser;
    use include_dir::{include_dir, Dir};
    use insta::assert_snapshot;
    use lustre_collector::parser::parse;
    use lustrefs_exporter::openmetrics::OpenTelemetryMetrics;
    use opentelemetry::metrics::MeterProvider;
    use prometheus::{Encoder as _, Registry, TextEncoder};
    use prometheus_parse::{Sample, Scrape};
    use std::{collections::HashSet, error::Error, fs};

    static VALID_FIXTURES: Dir<'_> =
        include_dir!("$CARGO_MANIFEST_DIR/../lustre-collector/src/fixtures/valid/");

    fn test_valid_fixtures_otel() {
        for dir in VALID_FIXTURES.find("*").unwrap() {
            match dir {
                include_dir::DirEntry::Dir(_) => {}
                include_dir::DirEntry::File(file) => {
                    let name = file.path().to_string_lossy();

                    let contents = file.contents_utf8().unwrap();

                    let result = parse()
                        .easy_parse(contents)
                        .map_err(|err| err.map_position(|p| p.translate_position(contents)))
                        .unwrap();

                    let (provider, registry) =
                        init_opentelemetry().expect("Failed to initialize OpenTelemetry");

                    let meter = provider.meter("lustre");

                    let otel = OpenTelemetryMetrics::new(meter);

                    crate::openmetrics::build_lustre_stats(&result.0, otel);

                    let mut buffer = vec![];
                    let encoder = TextEncoder::new();
                    let metric_families = registry.gather();
                    encoder.encode(&metric_families, &mut buffer).unwrap();

                    let x = String::from_utf8_lossy(&buffer).to_string();

                    assert_snapshot!(format!("valid_fixture_otel_{name}"), x);
                }
            }
        }
    }

    #[test]
    fn test_stats() {
        let output = include_str!("../fixtures/stats.json");

        let x = serde_json::from_str(output).unwrap();

        let (provider, registry) =
            init_opentelemetry().expect("Failed to initialize OpenTelemetry");

        let meter = provider.meter("lustre");

        let otel = OpenTelemetryMetrics::new(meter);

        crate::openmetrics::build_lustre_stats(&x, otel);

        insta::assert_snapshot!(get_output(&registry));
    }

    #[test]
    fn test_lnetctl_stats() {
        let output = include_str!("../fixtures/lnetctl_stats.json");

        let x = serde_json::from_str(output).unwrap();

        let (provider, registry) =
            init_opentelemetry().expect("Failed to initialize OpenTelemetry");

        let meter = provider.meter("lustre");

        let otel = OpenTelemetryMetrics::new(meter);

        crate::openmetrics::build_lustre_stats(&x, otel);

        insta::assert_snapshot!(get_output(&registry));
    }

    #[test]
    fn test_lnetctl_stats_mds() {
        let output = include_str!("../fixtures/stats_mds.json");

        let x = serde_json::from_str(output).unwrap();

        let (provider, registry) =
            init_opentelemetry().expect("Failed to initialize OpenTelemetry");

        let meter = provider.meter("lustre");

        let otel = OpenTelemetryMetrics::new(meter);

        crate::openmetrics::build_lustre_stats(&x, otel);

        insta::assert_snapshot!(get_output(&registry));
    }

    #[test]
    fn test_host_stats_non_healthy() {
        let output = include_str!("../fixtures/host_stats_non_healthy.json");

        let x = serde_json::from_str(output).unwrap();

        let (provider, registry) =
            init_opentelemetry().expect("Failed to initialize OpenTelemetry");

        let meter = provider.meter("lustre");

        let otel = OpenTelemetryMetrics::new(meter);

        crate::openmetrics::build_lustre_stats(&x, otel);

        insta::assert_snapshot!(get_output(&registry));
    }

    #[test]
    fn test_client_stats() {
        let output = include_str!("../fixtures/client.json");

        let x = serde_json::from_str(output).unwrap();

        let (provider, registry) =
            init_opentelemetry().expect("Failed to initialize OpenTelemetry");

        let meter = provider.meter("lustre");

        let otel = OpenTelemetryMetrics::new(meter);

        crate::openmetrics::build_lustre_stats(&x, otel);

        insta::assert_snapshot!(get_output(&registry));
    }
    use pretty_assertions::assert_eq;

    // Make sure metrics from the OpenTelemetry implementation are the same as the previous implementation
    #[tokio::test]
    async fn test_legacy_metrics() -> Result<(), Box<dyn std::error::Error>> {
        // Generate snapshots for the OpenTelemetry implementation
        test_valid_fixtures_otel();

        // Compare snapshots
        for dir in VALID_FIXTURES.find("*").unwrap() {
            match dir {
                include_dir::DirEntry::Dir(_) => {}
                include_dir::DirEntry::File(file) => {
                    let name = file.path().to_string_lossy().to_string().replace("/", "__");
                    println!("{}", format!("{}/src/snapshots/lustrefs_exporter__tests__valid_fixture_otel_{name}.snap", env!("CARGO_MANIFEST_DIR")));
                    let opentelemetry = read_metrics_from_snapshot(format!("{}/src/snapshots/lustrefs_exporter__tests__valid_fixture_otel_{name}.snap", env!("CARGO_MANIFEST_DIR")).as_str());
                    let previous_implementation = read_metrics_from_snapshot(
                        format!(
                            "{}/src/snapshots/lustrefs_exporter__tests__valid_fixture_{name}.snap",
                            env!("CARGO_MANIFEST_DIR")
                        )
                        .as_str(),
                    );
                    compare_metrics(&opentelemetry.unwrap(), &previous_implementation.unwrap());
                }
            }
        }
        Ok(())
    }

    fn read_metrics_from_snapshot(path: &str) -> Result<Scrape, Box<dyn Error>> {
        let content = fs::read_to_string(path)?;

        // Skip insta header
        let content = content
            .lines()
            .skip(4)
            .map(|s| Ok(s.to_owned()))
            .collect::<Vec<_>>();
        let parsed = Scrape::parse(content.into_iter())?;
        Ok(parsed)
    }

    fn normalize_sample(sample: &Sample) -> (String, Vec<(String, String)>, String) {
        let mut sorted_labels: Vec<_> = sample
            .labels
            .iter()
            .filter(|(k, _)| k != &&"otel_scope_name".to_string())
            .map(|(k, v)| (k.clone(), v.clone()))
            .collect();
        sorted_labels.sort();

        let value_str = match sample.value {
            prometheus_parse::Value::Counter(f) => f.to_string(),
            prometheus_parse::Value::Gauge(f) => f.to_string(),
            _ => "0.0".to_string(),
        };

        (sample.metric.clone(), sorted_labels, value_str)
    }

    fn normalize_docs(docs: &std::collections::HashMap<String, String>) -> Vec<(String, String)> {
        let mut sorted_docs: Vec<_> = docs
            .iter()
            .filter_map(|(k, v)| {
                if k != "target_info" {
                    Some((k.clone(), v.clone()))
                } else {
                    None
                }
            })
            .collect();
        sorted_docs.sort_by(|a, b| a.0.cmp(&b.0)); // Sort by key
        sorted_docs
    }

    fn compare_metrics(metrics1: &Scrape, metrics2: &Scrape) {
        // Skip OTEL specific metric
        let set1: HashSet<_> = metrics1
            .samples
            .iter()
            .filter(|s| s.metric != "target_info")
            .map(normalize_sample)
            .collect();
        let set2: HashSet<_> = metrics2
            .samples
            .iter()
            .filter(|s| s.metric != "target_info")
            .map(normalize_sample)
            .collect();

        let only_in_first: Vec<_> = set1.difference(&set2).collect();
        let only_in_second: Vec<_> = set2.difference(&set1).collect();

        let metric_value_comparison = if only_in_first.is_empty() && only_in_second.is_empty() {
            true
        } else {
            if !only_in_first.is_empty() {
                println!("Metrics only in first file:");
                for metric in only_in_first {
                    println!("{:?}", metric);
                }
            }
            if !only_in_second.is_empty() {
                println!("Metrics only in second file:");
                for metric in only_in_second {
                    println!("{:?}", metric);
                }
            }
            false
        };

        // Assert metrics values/labels are exactly the same
        assert!(
            metric_value_comparison,
            "Metrics values/labels are not the same"
        );

        // Normalize and compare metrics help
        let normalized_docs1 = normalize_docs(&metrics1.docs);
        let normalized_docs2 = normalize_docs(&metrics2.docs);

        assert_eq!(
            normalized_docs1, normalized_docs2,
            "Metrics help are not the same"
        );
    }

    fn get_output(registry: &Registry) -> String {
        let encoder = TextEncoder::new();
        let mut output = Vec::new();
        encoder.encode(&registry.gather(), &mut output).unwrap();
        String::from_utf8(output).unwrap()
    }
}
