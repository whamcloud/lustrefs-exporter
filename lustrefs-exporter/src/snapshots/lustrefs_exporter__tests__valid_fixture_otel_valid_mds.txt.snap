---
source: lustrefs-exporter/src/main.rs
expression: x
---
# HELP lustre_available_kilobytes Number of kilobytes readily available in the pool
# TYPE lustre_available_kilobytes gauge
lustre_available_kilobytes{component="mdt",target="ai400x2-MDT0000",otel_scope_name="lustre"} 405704300
lustre_available_kilobytes{component="ost",target="ai400x2-OST0000",otel_scope_name="lustre"} 32526506848
lustre_available_kilobytes{component="ost",target="ai400x2-OST0001",otel_scope_name="lustre"} 32532536272
# HELP lustre_capacity_kilobytes Capacity of the pool in kilobytes
# TYPE lustre_capacity_kilobytes gauge
lustre_capacity_kilobytes{component="mdt",target="ai400x2-MDT0000",otel_scope_name="lustre"} 414229176
lustre_capacity_kilobytes{component="ost",target="ai400x2-OST0000",otel_scope_name="lustre"} 34750424936
lustre_capacity_kilobytes{component="ost",target="ai400x2-OST0001",otel_scope_name="lustre"} 34750424936
# HELP lustre_connected_clients Number of connected clients
# TYPE lustre_connected_clients gauge
lustre_connected_clients{component="mdt",target="ai400x2-MDT0000",otel_scope_name="lustre"} 0
# HELP lustre_exports_dirty_total Total number of exports that have been marked dirty
# TYPE lustre_exports_dirty_total gauge
lustre_exports_dirty_total{component="ost",target="ai400x2-OST0000",otel_scope_name="lustre"} 0
lustre_exports_dirty_total{component="ost",target="ai400x2-OST0001",otel_scope_name="lustre"} 0
# HELP lustre_exports_granted_total Total number of exports that have been marked granted
# TYPE lustre_exports_granted_total gauge
lustre_exports_granted_total{component="ost",target="ai400x2-OST0000",otel_scope_name="lustre"} 278208
lustre_exports_granted_total{component="ost",target="ai400x2-OST0001",otel_scope_name="lustre"} 278208
# HELP lustre_exports_pending_total Total number of exports that have been marked pending
# TYPE lustre_exports_pending_total gauge
lustre_exports_pending_total{component="ost",target="ai400x2-OST0000",otel_scope_name="lustre"} 0
lustre_exports_pending_total{component="ost",target="ai400x2-OST0001",otel_scope_name="lustre"} 0
# HELP lustre_exports_total Total number of times the pool has been exported
# TYPE lustre_exports_total gauge
lustre_exports_total{component="mdt",target="ai400x2-MDT0000",otel_scope_name="lustre"} 15
lustre_exports_total{component="ost",target="ai400x2-OST0000",otel_scope_name="lustre"} 4
lustre_exports_total{component="ost",target="ai400x2-OST0001",otel_scope_name="lustre"} 4
# HELP lustre_free_kilobytes Number of kilobytes allocated to the pool
# TYPE lustre_free_kilobytes gauge
lustre_free_kilobytes{component="mdt",target="ai400x2-MDT0000",otel_scope_name="lustre"} 412951760
lustre_free_kilobytes{component="ost",target="ai400x2-OST0000",otel_scope_name="lustre"} 32878173676
lustre_free_kilobytes{component="ost",target="ai400x2-OST0001",otel_scope_name="lustre"} 32884203100
# HELP lustre_health_healthy Indicates whether the Lustre target is healthy or not. 1 is healthy, 0 is unhealthy.
# TYPE lustre_health_healthy gauge
lustre_health_healthy{otel_scope_name="lustre"} 1
# HELP lustre_inodes_free The number of inodes (objects) available
# TYPE lustre_inodes_free gauge
lustre_inodes_free{component="mdt",target="ai400x2-MDT0000",otel_scope_name="lustre"} 289693298
lustre_inodes_free{component="ost",target="ai400x2-OST0000",otel_scope_name="lustre"} 274719433
lustre_inodes_free{component="ost",target="ai400x2-OST0001",otel_scope_name="lustre"} 274719491
# HELP lustre_inodes_maximum The maximum number of inodes (objects) the filesystem can hold
# TYPE lustre_inodes_maximum gauge
lustre_inodes_maximum{component="mdt",target="ai400x2-MDT0000",otel_scope_name="lustre"} 289887952
lustre_inodes_maximum{component="ost",target="ai400x2-OST0000",otel_scope_name="lustre"} 274726912
lustre_inodes_maximum{component="ost",target="ai400x2-OST0001",otel_scope_name="lustre"} 274726912
# HELP lustre_ldlm_cbd_stats Gives information about LDLM Callback service.
# TYPE lustre_ldlm_cbd_stats counter
lustre_ldlm_cbd_stats{operation="ldlm_bl_callback",otel_scope_name="lustre"} 1
lustre_ldlm_cbd_stats{operation="req_active",otel_scope_name="lustre"} 1
lustre_ldlm_cbd_stats{operation="req_qdepth",otel_scope_name="lustre"} 1
lustre_ldlm_cbd_stats{operation="req_timeout",otel_scope_name="lustre"} 1
lustre_ldlm_cbd_stats{operation="req_waittime",otel_scope_name="lustre"} 1
lustre_ldlm_cbd_stats{operation="reqbuf_avail",otel_scope_name="lustre"} 3
# HELP lustre_lnet_mem_used Gives information about Lustre LNet memory usage.
# TYPE lustre_lnet_mem_used gauge
lustre_lnet_mem_used{otel_scope_name="lustre"} 62041660
# HELP lustre_lock_contended_total Number of contended locks
# TYPE lustre_lock_contended_total gauge
lustre_lock_contended_total{component="mdt",target="ai400x2-MDT0000",otel_scope_name="lustre"} 32
lustre_lock_contended_total{component="ost",target="ai400x2-OST0000",otel_scope_name="lustre"} 32
lustre_lock_contended_total{component="ost",target="ai400x2-OST0001",otel_scope_name="lustre"} 32
# HELP lustre_lock_contention_seconds_total Time in seconds during which locks were contended
# TYPE lustre_lock_contention_seconds_total gauge
lustre_lock_contention_seconds_total{component="mdt",target="ai400x2-MDT0000",otel_scope_name="lustre"} 2
lustre_lock_contention_seconds_total{component="ost",target="ai400x2-OST0000",otel_scope_name="lustre"} 2
lustre_lock_contention_seconds_total{component="ost",target="ai400x2-OST0001",otel_scope_name="lustre"} 2
# HELP lustre_lock_count_total Number of locks
# TYPE lustre_lock_count_total gauge
lustre_lock_count_total{component="mdt",target="ai400x2-MDT0000",otel_scope_name="lustre"} 0
lustre_lock_count_total{component="ost",target="ai400x2-OST0000",otel_scope_name="lustre"} 0
lustre_lock_count_total{component="ost",target="ai400x2-OST0001",otel_scope_name="lustre"} 0
# HELP lustre_lock_timeout_total Number of lock timeouts
# TYPE lustre_lock_timeout_total counter
lustre_lock_timeout_total{component="mdt",target="ai400x2-MDT0000",otel_scope_name="lustre"} 0
lustre_lock_timeout_total{component="ost",target="ai400x2-OST0000",otel_scope_name="lustre"} 0
lustre_lock_timeout_total{component="ost",target="ai400x2-OST0001",otel_scope_name="lustre"} 0
# HELP lustre_mds_mdt_fld_stats MDS mdt_fld stats
# TYPE lustre_mds_mdt_fld_stats gauge
lustre_mds_mdt_fld_stats{operation="fld_query",units="usecs",otel_scope_name="lustre"} 3
lustre_mds_mdt_fld_stats{operation="req_active",units="reqs",otel_scope_name="lustre"} 3
lustre_mds_mdt_fld_stats{operation="req_qdepth",units="reqs",otel_scope_name="lustre"} 3
lustre_mds_mdt_fld_stats{operation="req_timeout",units="secs",otel_scope_name="lustre"} 3
lustre_mds_mdt_fld_stats{operation="req_waittime",units="usecs",otel_scope_name="lustre"} 3
lustre_mds_mdt_fld_stats{operation="reqbuf_avail",units="bufs",otel_scope_name="lustre"} 7
# HELP lustre_mds_mdt_stats MDS mdt stats
# TYPE lustre_mds_mdt_stats gauge
lustre_mds_mdt_stats{operation="mds_connect",units="usecs",otel_scope_name="lustre"} 18
lustre_mds_mdt_stats{operation="obd_ping",units="usecs",otel_scope_name="lustre"} 1484
lustre_mds_mdt_stats{operation="req_active",units="reqs",otel_scope_name="lustre"} 1502
lustre_mds_mdt_stats{operation="req_qdepth",units="reqs",otel_scope_name="lustre"} 1502
lustre_mds_mdt_stats{operation="req_timeout",units="secs",otel_scope_name="lustre"} 1502
lustre_mds_mdt_stats{operation="req_waittime",units="usecs",otel_scope_name="lustre"} 1502
lustre_mds_mdt_stats{operation="reqbuf_avail",units="bufs",otel_scope_name="lustre"} 3571
# HELP lustre_mem_used Gives information about Lustre memory usage.
# TYPE lustre_mem_used gauge
lustre_mem_used{otel_scope_name="lustre"} 1457821523
# HELP lustre_mem_used_max Gives information about Lustre maximum memory usage.
# TYPE lustre_mem_used_max counter
lustre_mem_used_max{otel_scope_name="lustre"} 1458113635
# HELP lustre_oss_ost_create_stats OSS ost_create stats
# TYPE lustre_oss_ost_create_stats gauge
lustre_oss_ost_create_stats{operation="ost_statfs",units="usecs",otel_scope_name="lustre"} 4526
lustre_oss_ost_create_stats{operation="req_active",units="reqs",otel_scope_name="lustre"} 4526
lustre_oss_ost_create_stats{operation="req_qdepth",units="reqs",otel_scope_name="lustre"} 4526
lustre_oss_ost_create_stats{operation="req_timeout",units="secs",otel_scope_name="lustre"} 4526
lustre_oss_ost_create_stats{operation="req_waittime",units="usecs",otel_scope_name="lustre"} 4526
lustre_oss_ost_create_stats{operation="reqbuf_avail",units="bufs",otel_scope_name="lustre"} 9411
# HELP lustre_oss_ost_stats OSS ost stats
# TYPE lustre_oss_ost_stats gauge
lustre_oss_ost_stats{operation="obd_ping",units="usecs",otel_scope_name="lustre"} 10
lustre_oss_ost_stats{operation="ost_connect",units="usecs",otel_scope_name="lustre"} 28
lustre_oss_ost_stats{operation="ost_create",units="usecs",otel_scope_name="lustre"} 8
lustre_oss_ost_stats{operation="ost_get_info",units="usecs",otel_scope_name="lustre"} 4
lustre_oss_ost_stats{operation="req_active",units="reqs",otel_scope_name="lustre"} 50
lustre_oss_ost_stats{operation="req_qdepth",units="reqs",otel_scope_name="lustre"} 50
lustre_oss_ost_stats{operation="req_timeout",units="secs",otel_scope_name="lustre"} 50
lustre_oss_ost_stats{operation="req_waittime",units="usecs",otel_scope_name="lustre"} 50
lustre_oss_ost_stats{operation="reqbuf_avail",units="bufs",otel_scope_name="lustre"} 109
# HELP lustre_stats_total Number of operations the filesystem has performed.
# TYPE lustre_stats_total counter
lustre_stats_total{component="mdt",operation="statfs",target="ai400x2-MDT0000",otel_scope_name="lustre"} 1691
# HELP target_info Target metadata
# TYPE target_info gauge
target_info{service_name="lustrefs-exporter",telemetry_sdk_language="rust",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="0.29.0"} 1
